Zend/zend_language_scanner.l

```diff
@@ -2102,21 +2172,89 @@ inline_char_handler:
 	}
 
 	heredoc_label->label = estrndup(s, heredoc_label->length);
+	heredoc_label->indentation = 0;
+	saved_cursor = YYCURSOR;
+
+	zend_ptr_stack_push(&SCNG(heredoc_label_stack), (void *) heredoc_label);
+
+	while (YYCURSOR < YYLIMIT && (*YYCURSOR == ' ' || *YYCURSOR == '\t')) {
+		if (*YYCURSOR == '\t') {
+			spacing |= HEREDOC_USING_TABS;
+		} else {
+			spacing |= HEREDOC_USING_SPACES;
+		}
+		++YYCURSOR;
+	}
+
+	if (YYCURSOR == YYLIMIT) {
+		YYCURSOR = saved_cursor;
+		RETURN_TOKEN(T_START_HEREDOC);
+	}
 
 	/* Check for ending label on the next line */
 	if (heredoc_label->length < YYLIMIT - YYCURSOR && !memcmp(YYCURSOR, s, heredoc_label->length)) {
 		YYCTYPE *end = YYCURSOR + heredoc_label->length;
 
-		if (*end == ';') {
-			end++;
-		}
+		if (!IS_LABEL_START(*end)) {
+			if (spacing == (HEREDOC_USING_SPACES | HEREDOC_USING_TABS)) {
+				zend_throw_exception(zend_ce_parse_error, "Invalid indentation - tabs and spaces cannot be mixed", 0);
+				RETURN_TOKEN(T_ENCAPSED_AND_WHITESPACE);
+			}
+
+			if (*end == ';') {
+				end++;
+			}
 
-		if (*end == '\n' || *end == '\r') {
 			BEGIN(ST_END_HEREDOC);
+			is_heredoc = 0;
+		} else {
+			YYCURSOR = saved_cursor;
 		}
+	} else {
+		YYCURSOR = saved_cursor;
 	}
 
-	zend_ptr_stack_push(&SCNG(heredoc_label_stack), (void *) heredoc_label);
+	if (is_heredoc && !SCNG(heredoc_scan_ahead)) {
+		zend_lex_state current_state;
+		int heredoc_nesting_level = 1;
+
+		zend_save_lexical_state(&current_state);
+
+		SCNG(heredoc_scan_ahead) = 1;
+
+		zend_ptr_stack_reverse_apply(&current_state.heredoc_label_stack, copy_heredoc_label_stack);
+
+		while (heredoc_nesting_level) {
+			zval zv;
+			int retval;
+
+			ZVAL_UNDEF(&zv);
+			retval = lex_scan(&zv);
+			zval_dtor(&zv);
+
+			if (EG(exception)) {
+				zend_restore_lexical_state(&current_state);
+				RETURN_TOKEN(retval);
```

Comment:
I don't think returning `retval` here is correct. I think this exception should actually be discarded and detected at a later time.